File: Data Structures & Algorithms/three-integer-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result; 
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 2; i++) {
            if (i>0 && nums[i] == nums[i-1]) continue;
            int l = i+1;
            int r = nums.size() - 1;
            int target = -nums[i];
            while (l < r) {
                int curr = nums[l] + nums[r];
                if (target < curr) {
                    r--;
                } else if (curr < target) {
                    l++;
                } else {
                    vector<int> add_result = { nums[i], nums[l], nums[r] };
                    int found = 0; 
                    for(int j=0;j<result.size();j++) 
                        if (result[j] == add_result) found = 1;
                    if (!found) result.push_back(add_result);
                    l++;
                    r--;
                }
            }
        }
        return result;
    }
};

File: Data Structures & Algorithms/remove-duplicates-from-sorted-array/submission-1.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        # 1 2 2 3 3 4 4 5
        # 1 2 3 4 5 
        n = len(nums)
        if n < 2:
            return n

        l = 0
        r = 1

        while l < n and r < n:
            # print(l, r, nums[l], nums[r], nums)
            if nums[r] > nums[l]:
                l = r
                r += 1
            else:
                temp = r
                while r < n and nums[r] <= nums[l]:
                    # print(l, r)
                    r += 1
                # print(r)
                if r < n:
                    nums[temp] = nums[r]
                    l = temp 
                    r = temp + 1
                else: 
                    break 
        
        return l+1


